staging/src/k8s.io/legacy-cloud-providers/gce/gce.go

```diff
@@ -793,6 +793,29 @@ func (g *Cloud) HasClusterID() bool {
 	return true
 }
 
+// GetBasePath returns the compute API endpoint with the `projects/` element
+// compute API v0.36 changed basepath and dropped the `projects/` suffix, therefore suffix
+// must be added back when generating compute resource urls.
+func (g *Cloud) GetBasePath() string {
+	basePath := g.ComputeServices().GA.BasePath
+	return GetProjectBasePath(basePath)
```

Comment:
compute this once at Cloud construction time in CreateGCECloud, based on service.BasePath, and store in a private member field we can simply return here

Reply:
also, this code in CreateGCECloud doesn't look correct to me any more now that the go client library doesn't expect service.BasePath to end in `/projects/`:

```
	if config.APIEndpoint != "" {
		service.BasePath = fmt.Sprintf("%sprojects/", config.APIEndpoint)
		serviceBeta.BasePath = fmt.Sprintf("%sprojects/", strings.Replace(config.APIEndpoint, "v1", "beta", -1))
		serviceAlpha.BasePath = fmt.Sprintf("%sprojects/", strings.Replace(config.APIEndpoint, "v1", "alpha", -1))
	}
```

cc @freehan as author of that code in https://github.com/kubernetes/kubernetes/commit/3e8b4a27c48aae548234bef0a4e87c54e4a03ef2

I think the addition of the `/projects/` suffix should either be dropped or made conditional, like this:

```diff
diff --git a/staging/src/k8s.io/legacy-cloud-providers/gce/gce.go b/staging/src/k8s.io/legacy-cloud-providers/gce/gce.go
index 1a42cc0b0b2..8643f0a56f4 100644
--- a/staging/src/k8s.io/legacy-cloud-providers/gce/gce.go
+++ b/staging/src/k8s.io/legacy-cloud-providers/gce/gce.go
@@ -436,9 +436,19 @@ func CreateGCECloud(config *CloudConfig) (*Cloud, error) {
 	// For example,
 	// staging API endpoint: https://www.googleapis.com/compute/staging_v1/
 	if config.APIEndpoint != "" {
-		service.BasePath = fmt.Sprintf("%sprojects/", config.APIEndpoint)
-		serviceBeta.BasePath = fmt.Sprintf("%sprojects/", strings.Replace(config.APIEndpoint, "v1", "beta", -1))
-		serviceAlpha.BasePath = fmt.Sprintf("%sprojects/", strings.Replace(config.APIEndpoint, "v1", "alpha", -1))
+		computeServiceBasePath := config.APIEndpoint
+		if strings.HasSuffix(service.BasePath, "/projects/") {
+			// Only add a /projects/ suffix if the default compute service base path ends in /projects/
+			//
+			// The default library BasePath changed from 
+			// https://compute.googleapis.com/compute/v1/projects/ in v0.35.0 
+			// to 
+			// https://compute.googleapis.com/compute/v1/ in v0.36.0
+			computeServiceBasePath = getProjectsBasePath(computeServiceBasePath)
+		}
+		service.BasePath = computeServiceBasePath
+		serviceBeta.BasePath = strings.Replace(computeServiceBasePath, "v1", "beta", -1)
+		serviceAlpha.BasePath = strings.Replace(computeServiceBasePath, "v1", "alpha", -1)
 	}
 
 	containerService, err := container.NewService(context.Background(), option.WithTokenSource(config.TokenSource))
```

Reply:
I am thinking instead of trying to match the googleapis Basepath, we should always change the basepath here to what we expect the basepath to look like. I haven't run tests on this to verify that this behavior doesn't break something else, but I am envisioning something like below. 

```
computeServiceBasePath := service.BasePath
if config.APIEndpoint != "" {
    computeServiceBasePath := config.APIEndpoint
}
computeServiceBasePath = getProjectsBasePath(computeServiceBasePath)
service.BasePath = computeServiceBasePath
serviceBeta.BasePath = strings.Replace(computeServiceBasePath, "v1", "beta", -1)
serviceAlpha.BasePath = strings.Replace(computeServiceBasePath, "v1", "alpha", -1)
```

The benefit is all of the code that depends on this will be guaranteed to have the same view of the basepath. This feels less risky and more scalable than changing every usage of Basepath in the code. By changing here, all the dependent code can always operate on the assumption that `/projects/` will exist. In future breaking changes we only fix in one location. We can modify the existing unit test to be very explicit about where the fix needs to go too.

What do you think?
cc @prameshj, @freehan

Reply:
I don't know what the client library uses basepath for, but changing it to point to a different path level than the default does not seem like a good idea to me

Reply:
That is also a valid concern. My other suggestion is instead of altering the basepath keep an internal basepath variable on the Cloud object. We fix it here, and all the code would switch to using that instance variable instead of the basepath on the compute services resource.

Reply:
> My other suggestion is instead of altering the basepath keep an internal basepath variable on the Cloud object. We fix it here, and all the code would switch to using that instance variable instead of the basepath on the compute services resource.

that was my recommendation in https://github.com/kubernetes/kubernetes/pull/102913#discussion_r669671316

Reply:
Oh yes, sorry missed that comment. I agree with that recommendation.